cves/CVE-2014-3583.yml

```diff
@@ -223,4 +249,23 @@ mistakes:
     Use those questions to inspire your answer. Don't feel obligated to answer
     every one. Write a thoughtful entry here that those ing the software
     engineering industry would find interesting.
-  answer:
+  answer: |
+    This was a coding mistake, and to an extent, a design mistake. When Apache was
+    initially writing this function, they assumed that any header coming in from another
+    server would include the buffer end character, '\0'. This null character is the only
+    mechanism that stops C from reading memory past the end of the string. 
+
+    In C, every string allocated includes a null character at the end. However, the
+    vulnerability specifically relates to large headers coming specficially from a remote
+    FastCGI server. One good principal in secure coding is to not trust an input's validity
+    if it is coming from outside a trust boundary. In this case, the Apache developers
+    placed too much trust on the assumption that any header would include a null character.
+    It lead to them writing code that relied solely on the existence of an end character in input
+    to prevent overflow errors. The developers should have not trusted the validity of 
+    outside input and placed checks to prevent memory from being read
+
+    The fix is extremely simple - check the length of an incoming string and read only up to the 
+    length of that string. This is a good fix, since there isn't a reliance on outside input - the
+    headers are checked after they are received. Another potential fix Apache could consider
+    implementing is to simply limit the length of the headers, since this vulnerability specifically
+    appears on headers over a certain length.
```

Comment:
With the addition of new standards every day (especially years ago) the need for header length is growing larger and larger, if Apache had limited the size of the headers, would they have run into issues with support for newer standards? If so, would there be a better way to fix this?

Reply:
I may be biased, but it seems to me that this is a design flaw, if this is an issue in this specific module, how may other modules are also parsing HTTP request headers themselves, wouldn't it be better for there to be a module dedicated to parsing requests to turn the untrusted input into a trusted form?  I agree that at the face of it, this was a simple coding mistake, but some coding mistakes can be prevented with a solid design.

Reply:
> how may other modules are also parsing HTTP request headers themselves, wouldn't it be better for there to be a module dedicated to parsing requests to turn the untrusted input into a trusted form

I think this is a really good point and I'll add it into the YAML.